generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,328 @@
+#include "generator/camera_info_collector.hpp"
+
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+
+#include <algorithm>
+
+namespace generator
+{
+// TODO (@gmoryes) add "inline static ..." after moving to c++17
+uint32_t constexpr CamerasInfoCollector::kLatestVersion;
+double constexpr CamerasInfoCollector::Camera::kEqualityEps;
+double constexpr CamerasInfoCollector::Camera::kSignificantPartOfSegmentCoef;
+double constexpr CamerasInfoCollector::kMaxDistFromCameraToClosestSegmentMeters;
+double constexpr CamerasInfoCollector::kSearchCameraRadiusMeters;
+uint32_t constexpr CamerasInfoCollector::kMaxCameraSpeedKmpH;
+
+CamerasInfoCollector::CamerasInfoCollector(std::string const & dataFilePath,
+                                           std::string const & camerasInfoPath,
+                                           std::string const & osmIdsToFeatureIdsPath)
+{
+  std::map<osm::Id, uint32_t> osmIdToFeatureId;
+  if (!routing::ParseOsmIdToFeatureIdMapping(osmIdsToFeatureIdsPath, osmIdToFeatureId))
+  {
+    LOG(LWARNING, ("An error happened while parsing feature id to osm ids mapping from file:",
+                   osmIdsToFeatureIdsPath));
+
+    m_valid = false;
+    return;
+  }
+
+  if (!ParseIntermediateInfo(camerasInfoPath, osmIdToFeatureId))
+  {
+    LOG(LWARNING, ("Unable to parse intermediate file(", camerasInfoPath, ") about cameras info"));
+    m_valid = false;
+    return;
+  }
+
+  platform::LocalCountryFile file = platform::LocalCountryFile::MakeTemporary(dataFilePath);
+  auto registerResult = m_dataSource.RegisterMap(file);
+  if (registerResult.second != MwmSet::RegResult::Success)
+  {
+    LOG(LWARNING, ("Unable to RegisterMap:", dataFilePath));
+    m_valid = false;
+    return;
+  }
+
+  for (auto & camera : m_cameras)
+  {
+    camera.ParseDirection();
+    camera.FindClosestSegment(m_dataSource, registerResult.first);
+  }
+}
+
+void CamerasInfoCollector::Camera::FindClosestSegment(FrozenDataSource const & dataSource,
+                                                      MwmSet::MwmId const & mwmId)
+{
+  if (!m_ways.empty())
+  {
+    // If m_ways is not empty. It means, that one of point it is our camera.
+    // So we should find it in feature's points.
+    for (auto it = m_ways.begin(); it != m_ways.end();)
+    {
+      if (auto id = FindMyself(static_cast<uint32_t>(it->m_featureId), dataSource, mwmId))
+      {
+        it->m_segmentId = (*id).second;
+        it->m_coef = (*id).first;  // Camera starts at the begin or the end of segment.
+        CHECK(it->m_coef == 0.0 || it->m_coef == 1.0, ("Coefficient must be 0.0 or 1.0 here"));
+        ++it;
+      }
+      else
+      {
+        m_ways.erase(it);
```

Comment:
https://en.cppreference.com/w/cpp/container/vector/erase

Invalidates iterators and references at or after the point of the erase, including the end() iterator.